### PR TITLE
nbShards set to number of shards without meta and divideByZero fix in shuffler

### PR DESCRIPTION
**Pull request by raduchis**
nbShards was set to len(eligibleMap) and in fact it should contain just the number of shards without the meta
distribute nodes to waiting lists crashed with divideByZero if no validators were in the waitinglists 

1. fixed nbShards both in nodesCoordinator and in the shuffler
2. added lists in waitinglists if they do not exist

### CODE REVIEW

File: sharding/indexHashedNodesCoordinator.go [hunk 1]
```diff
@@ -637,12 +648,14 @@ func (ihgs *indexHashedNodesCoordinator) computeNodesConfigFromList(
 		sort.Sort(validatorList(leavingList))
 	}
 
+	nbShards := core.MaxInt(0, len(eligibleMap)-1)
+
```
Comment (sasurobert) [inline sharding/indexHashedNodesCoordinator.go:652]: Shouldn't error when nbShards==0?

File: sharding/errors.go [hunk 2]
```diff
@@ -118,5 +118,5 @@ var ErrNilOwnPublicKey = errors.New("nil own public key")
 // ErrNilEndOfProcessingHandler signals that a nil end of processing handler has been provided
 var ErrNilEndOfProcessingHandler = errors.New("nil end of processing handler")
 
-// ErrNilDestinationForDistribute signals that a nil value was provided for destination of distributedNodes
-var ErrNilDestinationForDistribute = errors.New("nil destination list for distributeNodes")
+// ErrNilorEmptyDestinationForDistribute signals that a nil or empty value was provided for destination of distributedNodes
```
Comment (SebastianMarian) [inline sharding/errors.go:121]: ErrNilOr instead ErrNilor

File: sharding/indexHashedNodesCoordinator_test.go [hunk 3]
```diff
@@ -1498,6 +1500,189 @@ func TestIndexHashedNodesCoordinator_ShuffleOutNotFound(t *testing.T) {
 	require.Equal(t, expectedShardForNotFound, newShard)
 }
 
+func TestIndexHashedNodesCoordinator_computeNodesConfigFromList_NoValidators(t *testing.T) {
+	t.Parallel()
```
Comment (SebastianMarian) [inline sharding/indexHashedNodesCoordinator_test.go:1504]: Add empty line

File: sharding/indexHashedNodesCoordinator_test.go [hunk 4]
```diff
@@ -1498,6 +1500,189 @@ func TestIndexHashedNodesCoordinator_ShuffleOutNotFound(t *testing.T) {
...
+	assert.Equal(t, 0, len(newNodesConfig.waitingMap))
+	assert.NotNil(t, newNodesConfig.leavingMap)
+	assert.Equal(t, 0, len(newNodesConfig.leavingMap))
+	assert.NotNil(t, newNodesConfig.newList)
+	assert.Equal(t, 0, len(newNodesConfig.newList))
+	assert.Equal(t, nbShards, newNodesConfig.nbShards)
+}
+
+func TestIndexHashedNodesCoordinator_computeNodesConfigFromList_NilPk(t *testing.T) {
+	t.Parallel()
```
Comment (SebastianMarian) [inline sharding/indexHashedNodesCoordinator_test.go:1541]: Add empty line

File: sharding/indexHashedNodesCoordinator_test.go [hunk 5]
```diff
@@ -1498,6 +1500,189 @@ func TestIndexHashedNodesCoordinator_ShuffleOutNotFound(t *testing.T) {
...
+
+	newNodesConfig, err := ihgs.computeNodesConfigFromList(validatorInfos)
+
+	assert.Nil(t, newNodesConfig)
+	assert.NotNil(t, err)
+	assert.Equal(t, ErrNilPubKey, err)
+}
+
+func TestIndexHashedNodesCoordinator_computeNodesConfigFromList_Validators(t *testing.T) {
+	t.Parallel()
```
Comment (SebastianMarian) [inline sharding/indexHashedNodesCoordinator_test.go:1573]: Add empty line

File: sharding/errors.go (hunk 2 above)
Comment (raduchis) [reply sharding/errors.go:121]: fixed

File: sharding/indexHashedNodesCoordinator_test.go (hunk 3 above)
Comment (raduchis) [reply sharding/indexHashedNodesCoordinator_test.go:1504]: fixed

File: sharding/indexHashedNodesCoordinator_test.go (hunk 4 above)
Comment (raduchis) [reply sharding/indexHashedNodesCoordinator_test.go:1541]: fixed

File: sharding/indexHashedNodesCoordinator_test.go (hunk 5 above)
Comment (raduchis) [reply sharding/indexHashedNodesCoordinator_test.go:1573]: fixed

File: sharding/indexHashedNodesCoordinator.go (hunk 1 above)
Comment (raduchis) [reply sharding/indexHashedNodesCoordinator.go:652]: you could still have enough in the waiting lists to fill all shards, maybe... but there is a problem with the state in the trie if there are no validators in the eligible map. I added some time ago a "// TODO: compare with previous nodesConfig if exists" in order to validate that what comes from the startOfEpochBlock matches what the validator had in the previous epoch. If the epochs are consecutive, there is a problem and maybe an error should occur.

File: sharding/indexHashedNodesCoordinator.go [hunk 6]
```diff
@@ -648,7 +648,11 @@ func (ihgs *indexHashedNodesCoordinator) computeNodesConfigFromList(
 		sort.Sort(validatorList(leavingList))
 	}
 
-	nbShards := core.MaxInt(0, len(eligibleMap)-1)
+	if len(eligibleMap) == 0 {
```
Comment (SebastianMarian) [inline sharding/indexHashedNodesCoordinator.go:651]: I think that len(eligibleMap) should be greater than 1 as we should have at least 1 shard beside metachain, but you can leave it like this for now.

**Pull request merged**